server/jetstream_cluster.go

```diff
@@ -2918,143 +2932,137 @@ func isControlHdr(hdr []byte) bool {
+					}
+					for _, entry := range entries {
+						_, _, op, buf, err = decodeBatchMsg(entry.Data[1:])
+						if err != nil {
+							mset.batchMu.Unlock()
+							panic(err.Error())
 						}
-						if counter.ops == 0 {
-							delete(mset.clusteredCounterTotal, subject)
+						if err = js.applyStreamMsgOp(mset, op, buf, isRecovering); err != nil {
+							mset.batchMu.Unlock()
+							// Make sure to return remaining entries to the pool on an error.
+							for _, nce := range mset.batchEntries[j:] {
+								nce.ReturnToPool()
+							}
+							return 0, err
 						}
 					}
-					mset.clMu.Unlock()
+					// Return the entry to the pool now.
+					bce.ReturnToPool()
 				}
-
-				// Clear expected per subject state after processing.
-				if mset.expectedPerSubjectSequence != nil {
-					mset.clMu.Lock()
-					if subj, found := mset.expectedPerSubjectSequence[lseq]; found {
-						delete(mset.expectedPerSubjectSequence, lseq)
-						delete(mset.expectedPerSubjectInProcess, subj)
-					}
-					mset.clMu.Unlock()
+				if len(mset.batchEntries) == 0 {
+					// Get within the same entry, but within the range of this batch.
+					entries = ce.Entries[mset.batchEntryStart : i+1]
+				} else {
+					// Get all entries up to and including the current one.
+					entries = ce.Entries[:i+1]
 				}
-
-				if err != nil {
-					if err == errLastSeqMismatch {
-
-						var state StreamState
-						mset.store.FastState(&state)
-
-						// If we have no msgs and the other side is delivering us a sequence past where we
-						// should be reset. This is possible if the other side has a stale snapshot and no longer
-						// has those messages. So compact and retry to reset.
-						if state.Msgs == 0 {
-							mset.store.Compact(lseq + 1)
-							// Retry
-							err = mset.processJetStreamMsg(subject, reply, hdr, msg, lseq, ts, mt, sourced)
-						}
-						// FIXME(dlc) - We could just run a catchup with a request defining the span between what we expected
-						// and what we got.
+				// Process remaining entries in the current entry.
+				for _, entry := range entries {
+					_, _, op, buf, err = decodeBatchMsg(entry.Data[1:])
+					if err != nil {
+						mset.batchMu.Unlock()
+						panic(err.Error())
 					}
-
-					// Only return in place if we are going to reset our stream or we are out of space, or we are closed.
-					if isClusterResetErr(err) || isOutOfSpaceErr(err) || err == errStreamClosed {
-						return err
+					if err = js.applyStreamMsgOp(mset, op, buf, isRecovering); err != nil {
+						mset.batchMu.Unlock()
+						return 0, err
 					}
-					s.Debugf("Apply stream entries for '%s > %s' got error processing message: %v",
-						mset.account(), mset.name(), err)
+				}
+				// Clear state, batch was successful.
+				batchActiveId = _EMPTY_
+				mset.clearBatchStateLocked()
+				mset.batchMu.Unlock()
+				continue
+			} else if batchActiveId != _EMPTY_ {
+				// If a batch is abandoned without a commit, reject it.
+				mset.batchMu.Lock()
```

Comment:
Keep repeating ourselves here, maybe these are functions?

Reply:
With the move into a separate `batchApply` struct (https://github.com/nats-io/nats-server/pull/7067/commits/ef5c2f4dcfe9d4af29a9d95b9b1933d881b721d3) could remove the `batchActiveId` variable that needed to be kept up-to-date.

Like mentioned here https://github.com/nats-io/nats-server/pull/7067#discussion_r2220043051, could only introduce a `rejectBatchState()` that's used in two placed, the lock needs to be held for longer in all other cases.